packages/storage-plus/src/snapshot/mod.rs

```diff
@@ -0,0 +1,377 @@
+        if count <= 1 {
+            self.checkpoints.remove(store, height.into());
+            Ok(())
+        } else {
+            self.checkpoints.save(store, height.into(), &(count - 1))
+        }
+    }
+}
+
+impl<'a, K, T> Snapshot<'a, K, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    K: PrimaryKey<'a> + Prefixer<'a>,
+{
+    /// should_checkpoint looks at the strategy and determines if we want to checkpoint
+    pub fn should_checkpoint(&self, store: &dyn Storage, k: &K) -> StdResult<bool> {
+        match self.strategy {
+            Strategy::EveryBlock => Ok(true),
+            Strategy::Never => Ok(false),
+            Strategy::Selected => self.should_checkpoint_selected(store, k),
+        }
+    }
+
+    /// this is just pulled out from above for the selected block
+    fn should_checkpoint_selected(&self, store: &dyn Storage, k: &K) -> StdResult<bool> {
+        // most recent checkpoint
+        let checkpoint = self
+            .checkpoints
+            .range(store, None, None, Order::Descending)
+            .next()
+            .transpose()?;
+        if let Some((height, _)) = checkpoint {
+            // any changelog for the given key since then?
+            let start = Bound::inclusive(U64Key::from(height));
+            let first = self
+                .changelog
+                .prefix(k.clone())
+                .range(store, Some(start), None, Order::Ascending)
+                .next()
+                .transpose()?;
+            if first.is_none() {
+                // there must be at least one open checkpoint and no changelog for the given height since then
+                return Ok(true);
+            }
+        }
+        // otherwise, we don't save this
+        Ok(false)
+    }
+
+    // If there is no checkpoint for that height, then we return StdError::NotFound
+    pub fn assert_checkpointed(&self, store: &dyn Storage, height: u64) -> StdResult<()> {
+        let has = match self.strategy {
+            Strategy::EveryBlock => true,
+            Strategy::Never => false,
+            Strategy::Selected => self.checkpoints.may_load(store, height.into())?.is_some(),
+        };
+        match has {
+            true => Ok(()),
+            false => Err(StdError::not_found("checkpoint")),
+        }
+    }
+
+    pub fn has_changelog(&self, store: &mut dyn Storage, key: K, height: u64) -> StdResult<bool> {
+        Ok(self
+            .changelog
+            .may_load(store, (key, U64Key::from(height)))?
+            .is_some())
+    }
+
+    pub fn write_changelog(
+        &self,
+        store: &mut dyn Storage,
+        key: K,
+        height: u64,
+        old: Option<T>,
+    ) -> StdResult<()> {
+        self.changelog
+            .save(store, (key, U64Key::from(height)), &ChangeSet { old })
+    }
+
+    // may_load_at_height reads historical data from given checkpoints.
+    // Only returns `Ok` if we have the data to be able to give the correct answer
```

Comment:
This comment is incorrect.

Returns `StdError::NotFound` if we have no checkpoint (and can give no data)
Returns `Ok(None)` if there is a checkpoint, but no cached data (no changes since the checkpoint, caller should query current state)
Return `Ok(Some(x))`if there is a checkpoint and data written to changelog, returning the state at that time